// @saucelabs/cypress-plugin/src/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.afterRunTestReport = exports.Region = void 0;
const reporter_1 = __importDefault(require("./reporter"));
const cli_table3_1 = __importDefault(require("cli-table3"));
const chalk_1 = __importDefault(require("chalk"));
const testcomposer_1 = require("@saucelabs/testcomposer");
Object.defineProperty(exports, "Region", { enumerable: true, get: function () { return testcomposer_1.Region; } });
let reporterInstance;
const reportedSpecs = [];
const accountIsSet = function () {
    return process.env.SAUCE_USERNAME && process.env.SAUCE_ACCESS_KEY;
};
const onBeforeRun = function (details) {
    if (!accountIsSet()) {
        return;
    }
    reporterInstance.cypressDetails = details;
};
const onAfterSpec = async function (spec, results) {
    if (!accountIsSet()) {
        return;
    }
    try {
        const job = await reporterInstance.reportSpec(results);
        reportedSpecs.push({
            name: spec.name,
            jobURL: job.url,
        });
        console.log(`Report created: ${job.url}`);
        await reporterInstance.reportTestRun(results, job.id);
    }
    catch (e) {
        if (e instanceof Error) {
            console.error(`Failed to report ${spec.name} to Sauce Labs:`, e.message);
        }
    }
};
const onAfterRun = function () {
    if (!accountIsSet() || reportedSpecs.length == 0) {
        return;
    }
    const table = new cli_table3_1.default({
        head: ['Spec', 'Sauce Labs job URL'],
        style: {
            head: [],
            'padding-left': 2,
            'padding-right': 2,
        },
        chars: {
            'top-mid': '',
            'top-left': '  ┌',
            'left': '  │',
            'left-mid': '  ├',
            'middle': '',
            'mid-mid': '',
            'right': '│',
            'bottom-mid': '',
            'bottom-left': '  └',
        }
    });
    for (const spec of reportedSpecs) {
        table.push([spec.name, spec.jobURL]);
    }
    console.log('\n');
    console.log(chalk_1.default['gray']('='.repeat(100)));
    console.log('\nJobs reported to Sauce Labs:\n');
    console.log(table.toString());
};
/**
 * Converts the results of a cypress run (the result from `after:run` or `cypress.run()`) to a sauce test report.
 *
 * @param results cypress run results, either from `after:run` or `cypress.run()`
 * @returns {TestRun}
 */
async function afterRunTestReport(results) {
    const rep = new reporter_1.default(undefined);
    const testResults = [];
    results.runs?.forEach(run => {
        testResults.push({ spec: run.spec, tests: run.tests, video: run.video });
    });
    return await rep.createSauceTestReport(testResults);
}
exports.afterRunTestReport = afterRunTestReport;
function default_1(on, config, opts) {
    reporterInstance = new reporter_1.default(undefined, opts);
    on('before:run', onBeforeRun);
    on('after:run', onAfterRun);
    on('after:spec', onAfterSpec);
    return config;
}
exports.default = default_1;
