// @saucelabs/testcomposer/src/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestComposer = exports.Region = void 0;
const axios_1 = __importDefault(require("axios"));
const form_data_1 = __importDefault(require("form-data"));
// The Sauce Labs region.
var Region;
(function (Region) {
    Region["USWest1"] = "us-west-1";
    Region["EUCentral1"] = "eu-central-1";
    Region["Staging"] = "staging";
})(Region = exports.Region || (exports.Region = {}));
const apiURLMap = new Map([
    [Region.USWest1, 'https://api.us-west-1.saucelabs.com/v1/testcomposer'],
    [Region.EUCentral1, 'https://api.eu-central-1.saucelabs.com/v1/testcomposer'],
    [Region.Staging, 'https://api.staging.saucelabs.net/v1/testcomposer']
]);
const appURLMap = new Map([
    [Region.USWest1, 'https://app.saucelabs.com'],
    [Region.EUCentral1, 'https://app.eu-central-1.saucelabs.com'],
    [Region.Staging, 'https://app.staging.saucelabs.net']
]);
class TestComposer {
    constructor(opts) {
        this.opts = opts;
        this.url = apiURLMap.get(opts.region) || Region.USWest1;
        this.requestConfig = { auth: { username: this.opts.username, password: this.opts.accessKey }, headers: opts.headers };
    }
    /**
     * Create a report in form of a job on Sauce Labs.
     * @param req Job metadata.
     */
    async createReport(req) {
        const resp = await axios_1.default.post(this.url + '/reports', req, this.requestConfig);
        const id = resp.data.ID;
        return { id: id, url: appURLMap.get(this.opts.region) + '/tests/' + id };
    }
    /**
     * Upload assets.
     * @param jobId The job ID the assets should be attached to.
     * @param assets The assets to upload.
     */
    async uploadAssets(jobId, assets) {
        const form = new form_data_1.default();
        for (const asset of assets) {
            form.append('file', asset.data, { filename: asset.filename });
        }
        const resp = await axios_1.default.put(this.url + `/jobs/${jobId}/assets`, form, this.requestConfig);
        return resp.data;
    }
}
exports.TestComposer = TestComposer;
